### ISSUE TITLE
Call for new maintainers

### ISSUE THREAD

**Issue by adam-lynch** [nwjs-community/nw-builder#327]
Would you like to become a maintainer and help out on this project? We'd really appreciate it.

![Puppy dog eyes](http://24.media.tumblr.com/tumblr_m5ooh7WuPL1rxk0pjo1_250.gif)

---

I'm going to spend a bit of time this week on getting the project back on track, specifically getting newer versions working again. 

I thought the biggest issue was that alpha/beta versions weren't supported (#275) but I had been too crazy busy lately to fix it. I've realised today that it's much worse than that.

Also, what do you think are the biggest issues preventing it to work with the latest versions? 
- Be smarter about available versions (& support alpha builds) #275
- Please Upgrade packages for node 6.0.0 #321
- Is there a problem with OSX 32-bit? I saw someone mention it's deprecated/gone.
- Is there anything else critical?

---

Pinging people who have been contributing through issues & PRs lately; @evshiron @joshterrill @craigbloodworth @atlantidezign @piysingh @vankasteelj @cgallarno @sqwk @colinbes @tomlandia  @Sytten @zergius-eggstream @YacheLee @ARH-Digital @jgknott @AzazelN28 @freakinside @mathzol @y-lohse @gabepaez @bastimeyer @technicallyjosh @trevorah

**Comment by vankasteelj** [comment]
Not sure I'll have time, but I 100% behind the idea, I held back upgrading nwjs on my personnal project because of nw-builder not being able to understand > 12.3

**Comment by evshiron** [comment]
@adam-lynch Greetings Adam :)

Glad to have you back to the `nwjs/nw-builder`.
When I was using `nw-builder`, I found it didn't work with any later versions, so I forked and hacked the `nw-builder` to work in my situation.
Later I found my hacks were somewhat ugly and as there was a need for a working `nw-builder`, I created another `nwjs-builder` (along with `nwjs-download`).

There are points why I stopped hacking over `nw-builder`:
* It uses `semver` and a list of files to determine which files should be copied (yes I am talking about `lib/platforms.js`). The list changes from version to version and if a new version comes out, it might refuse to work before contributors update the list (also the `semver`). And the list will keep growing as time goes.
  * So in `nwjs-builder`, I decide to use excluding instead of including.
* It doesn't use `http://nwjs.io/versions.json`, which you have asked for. And talking about sharing caches, the `nw-download` is out of maintaining as well.
  * That's why I created `nwjs-download`.
* OSX ia32 builds are simply dropped after nw.js 0.12.x. And `nw-builder` builds for it in default, which brings some complaints from the new users.
  * In `nwjs-download`, target platforms are parsed from `versions.json`, which solves the issue in another way.
* Someone says `nw-builder` doesn't provide editing Windows executable information. `nw-builder` should be using `winresourcer` to edit the icon, but not the information.
  * I use `node-rcedit` to edit the icon and most of the information.
* In later versions, i18n files are provided in OSX builds. Editing `Info.plist` will not change the display name as long as they are here.
  * So I ignore the i18n files in `nwjs-builder`.
* <del>Personally I don't like Promises, so I switch over when doing `nwjs-builder`.</del>

I am afraid there will be a lot of work to do, but it would be nice if `nw-builder` is able to back on track. I am glad to help if there is a plan (but I am busy with final exams ATM).

**Comment by bastimeyer** [comment]
Thanks for bringing this project back to life!
Also thanks for asking, but I don't want to be a maintainer. If there's anything to fix and I have the time, I will send a PR, like I have done a couple of times before in the past.

---

This is what I think needs to be done to resolve all *critical* issues:

1. Use the NWjs versions manifest at http://nwjs.io/versions.json instead of parsing http://dl.nwjs.io/
This fixes the issues of new versions (>=0.13.0) and dropped platforms (osx32).

2. Add support for `stable`, `beta` and `lts` channels and `normal` and `sdk` build flavors.
  - Set the default channel to stable and default flavor to sdk.
  - `build` and `run` should be able to have different channel/flavor selections (e.g. normal for build and sdk for run).

2. Include the whole archive content instead of defining a list of files to be used for building.
When the file contents changed after the v0.9.2 update, I submitted a PR for a selection of multiple semver ranges... But is this still necessary? We know the archive content and it should be safe to assume that all files are needed. This fixes issues for new releases in the future. Maybe implement an exclude list.

5. Upgrade dependencies, remove unneeded dependencies.
Fixes node>=6 issues.

This is what I don't like about the current versions manifest and should be changed by the NWjs devs:

- It should only be accessed via https.
- It should include hashsums.
- It should include direct download links, so we don't need to build the URL ourself.

There are also a few other issues, like file permissions on Linux and OSX, but these are rather optional...

**Comment by Sytten** [comment]
I currently dont have time to put on nw-builder since I'm the official maintainer of Facebook Messenger For Desktop. 
But I would very much like to see the official version work again.
I currently use the work of @evshiron in MFD.

**Comment by joshterrill** [comment]
@evshiron @bastimeyer I couldn't have said it better myself. If there's anything I can do to help, just let me know what you want me to work on. Maybe we should get a priority list going on what should be done first?

**Comment by ntzrmtthihu777** [comment]
I'm rather new to nw.js and so on, but I've ran into and hopefully properly figured some fixes between nw-builder and nw.js's newer versions, so I'd be willing to put in some work on it :)

**Comment by adam-lynch** [comment]
Thanks for the feedback! I'm going to make a proper (broken down) list of tasks. Not sure which to use:

- ZenHub.
- Huboard.
- Waffle.io.
- Trello.
- Plain old GitHub issues.

I've also created a [Gitter room](https://gitter.im/nwjs/nw-builder) for this project just in case.

**Comment by evshiron** [comment]
Just a suggestion. [GFM Task List](https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments) should be enough for a task list, it's public and developer-friendly.

* [x] Task1
* [ ] Task2
* [ ] Task3

**Comment by adam-lynch** [comment]
@evshiron yeah. I'd be ok with that. All of the other options are public / have a public option and every one but Trello actually maps issues to tasks, so if you complete a task or close the issue (or add a tag, etc.) it gets synced to the other side.

**Comment by adam-lynch** [comment]
OK so the top priority should be to get the existing functionality back in working order. So just patch version bumps to begin with before we add any additional feature/behaviour.

Note: All help is appreciated here; code, managing issues, helpful replies, anything. Feel free to chip in if you never have, you're more than welcome :+1:

---

- [ ] Add all existing issues to their relevant milestones.

- [ ] Add labels to all existing issues; `bug`, `enhancement`, `needinfo`, etc.


### 2.2.1 - Phoenix

I've created a [`phoenix` branch](https://github.com/nwjs/nw-builder/tree/phoenix) and a [milestone](https://github.com/nwjs/nw-builder/milestones/2.2.1%20-%20Phoenix) for these.

- [ ] Close any pull-requests that no longer match up with how we plan to do things or else get them refactored to suit.

- [ ] Don't run permissions tests if current platform is Windows. They'll never pass and we shouldn't have some tests that we accept will fail.

- [ ] Update/prune dependencies to fix node 6+ issues. (#321) 
- [ ] Use the new `versions.json` instead of hitting dl.nwjs.io. Ignore versions with pre-release/build metadata in the version to begin with (i.e. `-rc1`, `-alpha2`, etc). Since we want to make this backwards compatible, we should still use dl.nwjs.io, etc. for legacy versions. (#275) 
- [ ] Include the whole content of the archive. If possible anyway. Implement an exclusion list if necessary. This might just be part of the above task (#275)
- [ ] Warn that versions before `0.12.3` are deprecated if one is used to build.

- [ ] Patch release.
- [ ] Close all issues now resolved.


### 2.2.2 - Editing executable information.

I've created a [milestone](https://github.com/nwjs/nw-builder/milestones/2.2.2%20-%20Editing%20executable%20information) for these.

- [ ] Close any pull-requests that no longer match up with how we plan to do things or else get them refactored to suit.

- [ ] Editing Windows executable information.
- [ ] In later versions, i18n files are provided in OSX builds. Editing Info.plist will not change the display name as long as they are here. We need to ignore them.

- [ ] Patch release.
- [ ] Close all issues now resolved.


### 2.3.0 - Alpha/beta builds, etc.

- [ ] Close any pull-requests that no longer match up with how we plan to do things or else get them refactored to suit.

- [ ] Add the ability for the user to somehow use pre-release versions (i.e. `-beta2`). The user should still be able to say give me `>0.13.0` and it will grab the latest applicable version.

- [ ] Minor release.
- [ ] Close all issues now resolved.



### 2.4.0 - Flavours

- [ ] Close any pull-requests that no longer match up with how we plan to do things or else get them refactored to suit.

- [ ] Add support for build flavours. I'm not sure what the API for this should be like. At first I thought we could accept the semver string as we do now and also accept an array of flavours to filter by. So, `>=0.13.0` and `stable` for example. But in `versions.json`, only the latest stable and lts versions are listed (it's not listed on every stable / lts version). P.S. @bastimeyer I specifically asked for the `versions.json` to give us direct URLs but it was decided not to do that I guess.

- [ ] Minor release. 
- [ ] Close all issues now resolved.



---

Beyond this, we all have ideas but let's leave it at that for now.

**Comment by gpetrov** [comment]
may I add to @adam-lynch list:

# 2.2.2 - Editing executable information.

- [ ] allow executable to be renamed on the Mac and chmod +x it
- [ ] do Info.plist as well InfoPlist.strings edits on Mac
- [ ] improve the ZIP option on Windows to be 3-state:
     - [ ] zip = true -> zip and append to executable (as it is now)
     - [ ] zip = false -> copy files to same root (as it is now)
     - [ ] zip = 'package.nw' -> zip and add as separate file 'package.nw' in the root
- [ ] use RCEDIT (https://github.com/atom/node-rcedit) on Windows for easy executable version change on Windows

# 2.5 Sign executables

- [ ] Allow all nwjs files to be signed

**Comment by bastimeyer** [comment]
> allow executable to be renamed on the Mac and chmod +x it

It's not just the OSX executable(s) having wrong permissions (`700` instead of `755`), but also all of the Linux files. In the old releases `<=0.12.3`, all of them are `755` instead of just the executable itself. In the new releases, all of them are `700` (executable) or `600` (rest).
This is more of a NWjs issue rather than nw-builder, because the downloaded archives contain these permissions. This could still be fixed here in the nw-builder though, since there may be situations where the builder cache dir was manipulated after a download. But let's not discuss it in this thread, #289 is already opened...

> Allow all nwjs files to be signed

If we're already signing the executables, then we could also implement packaging support for a later milestone, maybe by using [fpm](https://github.com/jordansissel/fpm)... And also add support for creating an installer for the files on Windows by using something like [nsis](http://nsis.sourceforge.net).

All of this is ofc highly optional...

**Comment by evshiron** [comment]
> If we're already signing the executables, then we could also implement packaging support for a later milestone, maybe by using fpm... And also add support for creating an installer for the files on Windows by using something like nsis.

I will refuse this proposition if it's in `evshiron/nwjs-builder`.
We can provide some tutorials about how to do them in the Wiki but should not make it a feature which we will have to maintain in the future.
Adding features like this will increase project complexity. If there is further issue, will we need to dive into the manual of those tools we have never used before? Just leave it to users. Task runners are extremely suitable for these jobs.

**Comment by adam-lynch** [comment]
@gpetrov thanks for the suggestions. Any issues related to editing executable information / icons were added to the 2.2.2 milestone.

I think the permissions, zipping, and packaging/signing stuff could be worried about after the versions I've outlined are done.

Since @evshiron mentioned complexity... I think eventually this project should be a bit simpler and we could have other projects alongside it which may or may not use it. For example, the CLI could be a separate project which uses this one. I'm sure we all have interesting ideas for features and additional modules but let's get this stuff sorted first 👍

**Comment by adam-lynch** [comment]
@malko the forker agrees this is the way to go 👍 

---

@evshiron @ntzrmtthihu777 @joshterrill @Sytten @vankasteelj could you test #333 and let me know if I've broken anything?

**Comment by Sytten** [comment]
I will release my beta version and then I will try it :)

**Comment by Bartozzz** [comment]
I did my own version of `nw-builder` some time ago. Right now it is a quite big project with a user-interface and a simple-to-use binary. It allows users to compile their project for desktop platforms (using `NW.js`), some mobile platforms (only `Android` right now) and browsers. I am planning to add some support for others platforms like Nintendo devices, Xbox, PS4, etc.

I did some "plugins" (idk. exactly how to call it - for example native notifications with an unified API for all the platforms), and a global configuration for all the platforms (icons/splasharts/application informations).

I wonder if you might want to use it for the new `nw-builder` release as it also adds supports for others compilers like `Cordova` and `Electron`. If not, I want to ask you if I can try to sell it on my portfolio for a small amount of money? I did put a lot of work and effort into this application. My builder is written from scratch and has nothing to do with yours.

Here are some screenshots:

### Terminal:
<img width="534" alt="zrzut ekranu 2016-06-29 o 01 39 46" src="https://cloud.githubusercontent.com/assets/6741707/16436044/dfddc6e8-3d9a-11e6-9a83-a5cf25353111.png">

### Custom application:
<img width="586" alt="zrzut ekranu 2016-06-29 o 01 37 51" src="https://cloud.githubusercontent.com/assets/6741707/16435996/9af20eae-3d9a-11e6-95ad-8994cc3da86f.png">
<img width="557" alt="zrzut ekranu 2016-06-29 o 01 38 18" src="https://cloud.githubusercontent.com/assets/6741707/16436000/a01aeb26-3d9a-11e6-8d1e-4cceb79abaff.png">
<img width="553" alt="zrzut ekranu 2016-06-29 o 01 38 32" src="https://cloud.githubusercontent.com/assets/6741707/16436005/a8fa50b0-3d9a-11e6-8ef2-d128469700a6.png">
<img width="565" alt="zrzut ekranu 2016-06-29 o 01 38 48" src="https://cloud.githubusercontent.com/assets/6741707/16436008/ab4133b6-3d9a-11e6-9edc-3554e508ccec.png">
<img width="860" alt="zrzut ekranu 2016-06-29 o 01 40 35" src="https://cloud.githubusercontent.com/assets/6741707/16436010/accf6964-3d9a-11e6-8496-343452c1eb48.png">
<img width="341" alt="zrzut ekranu 2016-06-29 o 01 39 19" src="https://cloud.githubusercontent.com/assets/6741707/16436011/ae655fd6-3d9a-11e6-8e40-abfa095da618.png">

**Comment by Sytten** [comment]
Dam this looks good!

**Comment by vankasteelj** [comment]
that's good for ppl wanting instructions, but what about projects using a server to auto build their app? We need programmable builds in one command line, to embed in gulp and that can then be used hooked when the main repo is updated

**Comment by Bartozzz** [comment]
It also provides a simple way to do it without the command line binary/dedicated application, very similar to `nw-builder`:

```javascript
import Build from "...";

const Builder = new Build( options );

Builder.on( "end",   archive => console.log( `Compiled! Output: ${archive}` ) );
Builder.on( "error", message => console.log( `Error! ${message}` ) );
Builder.build();
```

I could write a `gulp`/`grunt`/whatever plugin or even make some modifications to [`grunt-nw-builder](https://github.com/nwjs/grunt-nw-builder) plugin and it should work fine. As I said, this is a work in progress and I focused more on the application, mobile platforms and custom plugins.

**Comment by vankasteelj** [comment]
Then I didn't say a thing :p keep up the good work!

**Comment by adam-lynch** [comment]
@Bartozzz that looks good! You should definitely put it up online somewhere (and even open-source it 😄). I'd like to give it a try. But we're focusing right now on getting this project back to working (for newer versions) like it used to for older versions. We don't need to re-do a lot of the code. I think we're really close if you see #333. Thanks anyway though.

**Comment by adam-lynch** [comment]
OK, 2.2.1 is released which supports newer versions, alpha/beta builds, and fixes a few other little things. See #337 and #333. 

The only flavour supported right now is the SDK one (because it's most like older versions). We have a pull-request (#336) which adds the ability to choose flavours.

**Comment by rogerwang** [comment]
Do we have any new maintainer now? When will this repo enter the unmaintained state if no one is found?


CC @ghostoy

**Comment by joshterrill** [comment]
@Bartozzz that looks so good!

**Comment by Bartozzz** [comment]
@joshterrill Thanks a lot! There's a demo video of this project: https://www.youtube.com/watch?v=Ol9x73k48c4 - I will publish it as soon as possible [here](https://github.com/VulcanusProject) (I don't have much time right now).

**Comment by adam-lynch** [comment]
I've spoken to Roger directly in case as anyone is wondering 😄. I'm going to maintain this still but I can't do it on my own.

**Comment by matthew-dean** [comment]
What about a Patreon for nw-builder (or nw.js in general) for a dedicated maintainer like Vue.js did? https://www.patreon.com/evanyou

There's no reason it has to be entirely voluntary, and that might allow some room to get some end-user tools developed like actual distributables, since both nw.js and nw-builder fall just shy of installable packages. Or things like file associations. Or an integrated self-updater. I'd definitely kick in for someone's monthly salary.

**Comment by matthew-dean** [comment]
From Vue's Patreon: "If you run a business and is using Vue in a revenue-generating product, it would make business sense to sponsor Vue development: it ensures the project that your product relies on stays healthy and actively maintained. It can also help your exposure in the Vue community and makes it easier to attract Vue developers."

There are definitely a lot of businesses using NW.js and nw-builder. I was just hired to a place partly because of my NW.js experience. Let businesses invest in NW.js's success.

**Comment by adam-lynch** [comment]
@matthew-dean interesting. The main NW.js project used to use https://www.bountysource.com/ or something similar. @rogerwang would the right person to weigh in here.

**Comment by adam-lynch** [comment]
Big shoutout to @DblK and @langco. Between the two of them, we now (3.1.0) have support for selecting NW.js flavors other than `sdk`. 👏

**Comment by kongyuan** [comment]
@Bartozzz Your idea is great! How is it going for now? I can't open https://github.com/VulcanusProject

**Comment by tharatau** [comment]
@Bartozzz Did you ever end up release your version of `nw-builder`?

**Comment by Bartozzz** [comment]
@tharatau unfortunately not – I was not satisfied with the code-quality (I was just learning React at the time) and the performance, so I decided to postpone the release and it never happened. This project of mine is discontinued.

**Comment by tharatau** [comment]
Okay thanks for letting me know.
